### ISSUE TITLE
[P2P Distance] Create UpdateMoneyRequestDistanceRate

### ISSUE THREAD

**Issue by neil-marcellini** [Expensify/App#36969]
Create UpdateMoneyRequestDistanceRate API command in Web-E and Auth following [this plan](https://docs.google.com/document/d/1MepSFJHmETFRvT7KVlsKp-e5NlsS0Gn4P6cyVEPxIYg/edit#bookmark=id.kiyvn73dn3k6)
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0116944cbb002d044c</li>
        <li>Upwork Job ID: 1760074735350415360</li>
        <li>Last Price Increase: 2024-02-20</li>
    </ul>
</details>

**Comment by mollfpr** [comment]
Not overdue

**Comment by Beamanator** [comment]
Hey hey I'm happy to take this one one - do we have a dealine for implementing this bad boi?

**Comment by neil-marcellini** [comment]
Woah I totally missed that you were working on this! Thanks so much! I'm happy to help review the rest of it. Please DM me when it's ready for review of if you have any questions.

**Comment by neil-marcellini** [comment]
> I'm trying to figure out how to get an existing $distance amount from an existing transaction

Ah yeah that part wasn't described super clearly. You can get the existing distance from the transaction NVPs. Something like this.

```php
$transaction->getNVPs()["customUnit"]["quantity"];
```

**Comment by Beamanator** [comment]
Thanks @neil-marcellini ! Feel free to post-review the Auth PR if you're interested, but I'm actually OOO tomorrow so won't be continuing the Web-E changes till Monday! Here's my WIP PR for web-e: https://github.com/Expensify/Web-Expensify/pull/41495

And thanks for explaining where to get the distance!! I haven't worked much with distance requests, and it definitely isn't immediately obvious that it's in the quantity of the custom unit, so thanks for clarifying!! 🙏 hopefully I'll have that PR ready for review on Monday 🙏

**Comment by neil-marcellini** [comment]
Yep, check out [[$500] [P2P Distance] Create a new Rate field](https://github.com/Expensify/App/issues/36985). There's a PR up now which might be complete enough for testing [[App] Feat/36985 create new rate field](https://github.com/Expensify/App/pull/38543).

**Comment by Beamanator** [comment]
aah nice, it also looks like the "Edit rate" functionality will be added in [[$500] [HOLD 36985] [P2P Distance] Create UpdateMoneyRequestRate in App](https://github.com/Expensify/App/issues/36987#top), so I mayyyy wait to test with the changes made there?

**Comment by Beamanator** [comment]
@paultsimura the payload params we're expecting on `UpdateDistanceRequestRate` are as follows:

> (string $authToken, string $transactionID, string $customUnitRateID, int $optimisticReportActionID = 0)

**Comment by Beamanator** [comment]
Aah that's totally fine if we name the report action ID param `reportActionID`, in our backend I'll just quickly rename it to `optimistic*` so it's clear we're in optimistic land :D

**Comment by Beamanator** [comment]
Nice, sounds good - let me know if you need help but feel free to work on that fix in your PR 👍 🙏

**Comment by Beamanator** [comment]
Eek, I'm moving so slow here b/c it's a bit confusing how to break a new `computeDistanceCustomUnit` function out of `computeRouteAndCustomUnit` in the most efficient way possible 😅

**Comment by Beamanator** [comment]
@neil-marcellini question for you, which I asked in another issue, here: https://github.com/Expensify/App/issues/31673#issuecomment-2051719369

This command `UpdateMoneyRequestRate` will be used for P2P AND for money requests in workspaces, right? If so, I might have to deal with tax updates for workspace money requests, eh?

**Comment by Beamanator** [comment]
FYI I couldn't quite get this one ready by my EOD - I had to run out a bit early, I'll either be back late this evening or over the weekend for a few hours to make up for this.

@neil-marcellini (sorry to keep pinging you) - If you have time today, it would be great if you could give https://github.com/Expensify/Web-Expensify/pull/41495 a preliminary review -> Like to see if I'm going a decent direction that you were planning in your doc. It looked easyish and straightforwardish, but currently it's breaking `TrackExpenseTest` and there's some psalm errors which, for the life of me, i can't figure out why they don't exist in `main` 😅

**Comment by neil-marcellini** [comment]
> @neil-marcellini question for you, which I asked in another issue, here: [#31673 (comment)](https://github.com/Expensify/App/issues/31673#issuecomment-2051719369)
> 
> This command `UpdateMoneyRequestRate` will be used for P2P AND for money requests in workspaces, right? If so, I might have to deal with tax updates for workspace money requests, eh?

Only workspace requests. I answered in detail over on that issue [here](https://github.com/Expensify/App/issues/31673#issuecomment-2052308840).

**Comment by Beamanator** [comment]
Thanks @neil-marcellini 🙏 

Ok so it seeeeeems like we ARE going to have to deal with tax stuff possibly, but will cross that bridge in a bit 😅 

For now, I'm going to have to rewrite a few things, due to us agreeing ([here](https://github.com/Expensify/Web-Expensify/pull/41495#discussion_r1563048699)) that the current implementation isn't 1:1:1. Here's my new plan of action for now:

Web-E:
1. Continue accepting `authToken`, `transactionID`, `customUnitRateID`, `reportActionID` params
2. Since we don't want to do a Read call to get the transaction ID in PHP, we can't check if the transaction is a custom unit distance, AND we can't get the linked policy chat report, AND we can't get the transaction's policy data
3. So we'll pass the params directly to Auth to validate & do all the updates
4. We also can't do `TransactionUtils::pushTransactionUpdate` at the end, since this reads `iouReportID`, `iouThreadReportID`, both report's actions, their NVPs, and possibly more

Auth:
1. Do all the things Web-E was going in point 2 above in Peek
2. Also in Peek, Compute the custom distance amount & merchant, and custom unit info (from `TransactionAPI::getP2PDistanceRequestRate` or `TransactionAPI::getDistanceRequestRate`)
3. Use those bits ^ in the currently existing `auth.moneyRequest.updateRate` function
4. Figure out how to send all onyx updates via `queueOnyxUpdates(<report channel>, <reportID>, <updates>)

**Comment by neil-marcellini** [comment]
Thanks Alex, that plan sounds pretty good. I think you'll be happy to hear that my top focus right now is this [Migrate `TransactionUtils::pushTransactionUpdate` function to auth](https://github.com/Expensify/Expensify/issues/384098). Once that's done you should be able to use it in the new UpdateMoneyRequestRate Auth command.

I'm tackling that migration in two parts. In the first part I'll move most of it to Auth, and leave a small part in PHP to send the modified expense report action and the violations, because it's a massive amount of work to move those over. I'm working on a plan to distribute that migration work across the team, which will be the final part of moving pushTransactionUpdate to Auth.

**Comment by Beamanator** [comment]
Oh siiiick, that's great that we're taking that in chunks and i won't have to think about that part 😅 I'll get started on other pieces of the plan tomorrow, taking into consideration that i won't have to deal with migrating `pushTransactionUpdate` 👍 🙏

**Comment by Beamanator** [comment]
I like this plan, I can also make this change in the Auth command 👍

**Comment by Beamanator** [comment]
Web-e PR is up and waiting on Auth PR

Auth PR is slowly in progress, but is waiting on at least 2 other PRs most likely (Monil's PR for some Tax updates in Auth & Neil's PR for moving `pushTransactionUpdate` to Auth

**Comment by Beamanator** [comment]
I should be able to get back into this tomorrow, though i still believe i won't be able to complete this until the other tasks I mentioned are done - I'll try to find those WIP PRs to link here

**Comment by Beamanator** [comment]
Tax stuff: https://github.com/Expensify/Auth/pull/10517

**Comment by Beamanator** [comment]
Ooh ok some updates:
- `ForexUtils::format` was moved to `Currency::format`, which had some mega refactoring in https://github.com/Expensify/Auth/pull/10625 (merged)
- `Transaction::generateTaxUDF` was built in https://github.com/Expensify/Auth/pull/10517 (merged)
- `Transaction::pushTransactionUpdate` still being implemented in https://github.com/Expensify/Auth/pull/10487

**Comment by Beamanator** [comment]
Alright @neil-marcellini  - my Auth PR https://github.com/Expensify/Auth/pull/10638 is feelin much better - I haven't added in the stuff from ^ above PRs, but it's at least doing some of the web-e stuff... Do you mind taking a quick look to give any thoughts on where it's at for now?

**Comment by Beamanator** [comment]
Thanks for the initial review! Good questions, I got a few resolved but didn't have enough time (half day today & tomorrow) so will try to move forward tomorrow or early next week

**Comment by Beamanator** [comment]
WIP! Currently working on updating Currency::format to be able to format doubles as well, since currently it only accepts ints

**Comment by Beamanator** [comment]
FYI https://github.com/Expensify/Auth/pull/10809 just got merged, which unblocks https://github.com/Expensify/Auth/pull/10638 - hoping to get that one finished this week, as i believe all the other PRs it was held on are merged!

**Comment by Beamanator** [comment]
Made a bit of progress adding a few more tests today, didn't have much time w/ deploy duty

**Comment by Beamanator** [comment]
Ok @neil-marcellini I think I made a lot of progress on the PR today - I added `auth.transaction.pushTransactionUpdate` and related tests, fixed a few things, and made sure currency doubling looks good.

Mind giving it another text in the next few days? I'm probably going to have to chat with @MonilBhavsar to figure out what kind of tax stuff i may have to incorporate as well

**Comment by Beamanator** [comment]
Had to prioritize some API optimizations today, hoping to get back to this wednesday

**Comment by Beamanator** [comment]
tryiingggg to get this done by EOW

**Comment by Beamanator** [comment]
@MonilBhavsar is planning to help with the tax changes early this week, since tax stuff is confusing a.f.

**Comment by Beamanator** [comment]
Commenting again to hopefully remove the Overdue label

**Comment by neil-marcellini** [comment]
Waiting for updates from @MonilBhavsar on the PR

**Comment by Beamanator** [comment]
Auth PR was merged!! 🚀  I'm OOO this week but will get to the web-e changes to make sure it works well once I'm back!

**Comment by Beamanator** [comment]
ermagerd all PRs are finally merged! And web-e is on staging, so we're almost done with this!

**Comment by Beamanator** [comment]
Web-E Went to prod 4 days ago! @neil-marcellini want me to close this out? ORr should we keep it open while the App PR is being tested w/ all the latest code?

**Comment by neil-marcellini** [comment]
@Beamanator nah we can close this, there's a separate App issue to implement the front end part. You're work here is done! Unless of course Pavlo has found a legitimate bug 😬 😆 

[[$500][P2P Distance] Create UpdateMoneyRequestDistanceRate in App](https://github.com/Expensify/App/issues/36987#top)

**Comment by Beamanator** [comment]
Yo @paultsimura I'd be happy to help look into this, could you give me more specific test steps so I can try to reproduce & figure out exactly what you're seeing? :D

**Comment by Beamanator** [comment]
Nice, ok so there's definitely a bug, it LOOKS like the problem is that some C++ code isn't accurately converting an integer value into a float 😅

**Comment by Beamanator** [comment]
Ok ya so the problem

**Comment by Beamanator** [comment]
@paultsimura Most likely tomorrow, as long as it gets reviewed and merged today

**Comment by Beamanator** [comment]
Seems a lot of people were off at the end of last week for the 4th of July, so the changes have not been merged yet 🙏

**Comment by Beamanator** [comment]
Fix got to Prod end of last week! 🚀
